proposals/infra/INF-0004-binding-resource-binding-diagnostics.md

```diff
@@ -0,0 +1,212 @@
+| CBuffer | b |  "Object type '%0' with resource class "CBuffer" expects binding prefix 'b', but binding prefix '%1' was given." |
+
+There are certain resource kinds within the SRV resource
+class that are "sampleable". These resources are typically
+prefixed with "Texture" in actual HLSL source. So, there will
+be a special check to see whether the candidate resource is 
+prefixed with "Texture", and if so, the "s" binding prefix will
+also be permitted for that resource.
+
+If the given candidate resource is a user-defined type (UDT), then further
+analysis is necessary. The first step is to gather all registers that
+are being bound to this declaration, and collect the binding prefixes
+that are being specified. The UDT must have at least one valid resource
+that can be bound to the provided binding prefix. If not, an error must be
+emitted stating that "No resource contained in struct '%0' can be bound
+to binding prefix '%1'". There are no issues if a UDT has more resources
+than there are register binding statements, the resources will be bound to
+the next available space automatically, and so compilation can succeed.
+Additionally, a singular resource can be bound multiple times to different
+spaces, so multiple binding prefixes are allowed for a single resource.
+Below is a table that represents the different behaviors that could arise
+given some examples of different UDT's:
+
+| UDT | Binding statements | Diagnostic |
+|-|-|-|
+| struct Foo {<br>  float f;<br>  Buffer<float> Buf;<br>  RWBuffer<float> RWBuf;<br>};| register(t0) : register(s0); | None. Success because Buf gets bound to t0 and RWBuf gets bound to s0, and f is skipped. 
+| struct Foo {<br>  float f;<br>  Buffer<float> Buf;<br>  RWBuffer<float> RWBuf;<br>  RWBuffer<float> RWBuf2;<br>};<br> | register(t0) : register(s0); | None. Success because Buf gets bound to t0  and RWBuf gets bound to s0, and f is skipped. RWBuf2 gets assigned to s1  even though there is no explicit binding for s1. |
+| struct Foo {<br>  float f;<br>  Buffer<float> Buf;<br>}; <br>| register(t0) : register(s0); | None. Success because Buf gets bound to t0. Buf will also be bound to s0.|
+| struct Foo {<br>  struct Bar {<br>    RWBuffer<int> a;<br>    };<br>    Bar b;<br>}; | register(t0) | None. Success because Bar, the struct within Foo, has a valid resource that can be bound to t0. |
+| struct Foo {<br>  float f;<br>}; <br>| register(t0) | DefaultError warning. "UDT resource 'Foo' does not contain an applicable resource type for binding prefix 't'"|
+| struct Foo {<br>  struct Bar {<br>      float f;<br>    }<br>    Bar b;<br>};<br> | register(t0) | DefaultError warning. "UDT resource 'Foo' does not contain an applicable resource type for binding prefix 't'"|
+
+Finally, if the candidate type is not a valid resource type or not a
+UDT, the final case will be entered. If the resource is among any
+numeric or non-intangible types, there is only one exception where it can

```

Comment:
> non-intangible

This double negative is confusing. I think that this is also the first occurance of "intangible" in the document. As this is a newly invented term (and possibly one that might be renamed in the language spec based on some other discussions I've heard), it'd be great to have something in here explaining what it means in the context of this document.

Reply:
I took some extra time to define intangible.